File: POJ/1187/main.cpp
#include <iostream>
#include <cstdio>
#include <memory.h>
using namespace std;
const int maxn = 12;
const int maxm = 32;
const int mod = 11380;
int dp[maxn][maxn][maxn][maxm];
bool vis[maxn][maxn][maxn][maxm];
int a,b,c,d;

int dfs(int l1,int l2,int l3,int dep)
{
  if(l1 == 0 && l2 == 0 && l3 == 0)
    {
      vis[l1][l2][l3][dep] = true;
      return dp[l1][l2][l3][dep] = 1;
    }
  if(dep == 0)
    {
      vis[l1][l2][l3][dep] = true;
      return dp[l1][l2][l3][dep] = 0;
    }
  if(vis[l1][l2][l3][dep])
    {
      return dp[l1][l2][l3][dep];
    }
  int res = 0;
  for(int i=0;i<=l3;i++)
    {
      if(i)
        {
          res += (dfs(0,0,i-1,dep-1) * dfs(l1,l2,l3-i,dep)) % mod;
          res %= mod;
        }
      for(int j=0;j<=l2;j++)
        {
          if(j)
            {
              res += (dfs(0,j-1,i,dep-1) * dfs(l1,l2-j,l3-i,dep)) % mod;
              res %= mod;
            }
          for(int k=1;k<=l1;k++)
            {
              res += (dfs(k-1,j,i,dep-1) * dfs(l1-k,l2-j,l3-i,dep)) % mod;
              res %= mod;
            }
        }
    }
  vis[l1][l2][l3][dep] = true;
  return dp[l1][l2][l3][dep] = res;
}

int main()
{
  memset(vis,false,sizeof(vis));
  while(~scanf("%d %d %d %d",&a,&b,&c,&d))
    {
      dfs(a,b,c,d);
      if(d) dfs(a,b,c,d-1);
      int ans;
      if(d == 0) ans = dp[a][b][c][d];
      else ans = ((dp[a][b][c][d] - dp[a][b][c][d-1]) % mod + mod) % mod;
      printf("%d\n",ans);
    }
  return 0;
}

File: POJ/1187/Main.java
import java.util.Scanner;

public class Main {
    private static int[][][][] dp;
    private static boolean[][][][] vis;
    private static final int MOD = 11380;

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        int l1 = scan.nextInt();
        int l2 = scan.nextInt();
        int l3 = scan.nextInt();
        int d = scan.nextInt();

        dp = new int[l1+1][l2+1][l3+1][d+1];
        vis = new boolean[l1+1][l2+1][l3+1][d+1];

        for (int i = 0; i <= l1; i++) {
            for (int j = 0; j <= l2; j++) {
                for (int k = 0; k <= l3; k++) {
                    dp[i][j][k][0] = 0;
                    vis[i][j][k][0] = true;
                }
            }
        }

        for (int i = 0; i <= d; i++) {
            dp[0][0][0][i] = 1;
            vis[0][0][0][i] = true;
        }

        if (d == 0) {
            System.out.println(dp[l1][l2][l3][d]);
        } else {
            int ans = dfs(l1, l2, l3, d) - dfs(l1, l2, l3, d-1);
            while (ans < 0) ans += MOD;
            System.out.println(ans % MOD);
        }
    }

    private static int dfs(int l, int m, int n, int d) {
        if (vis[l][m][n][d]) return dp[l][m][n][d];
        int add = 0;
        for (int nn = 0; nn < n; nn++) {
            add += (dfs(0, 0, nn, d-1) * dfs(l, m, n - nn - 1, d)) % MOD;
            add %= MOD;
        }
        for (int mm = 0; mm < m; mm++) {
            for (int nn = 0; nn <= n; nn++) {
                add += (dfs(0, mm, nn, d-1) * dfs(l, m - mm - 1, n - nn, d)) % MOD;
                add %= MOD;
            }
        }
        for (int ll = 0; ll < l; ll++) {
            for (int mm = 0; mm <= m; mm++) {
                for (int nn = 0; nn <= n; nn++) {
                    add += (dfs(ll, mm, nn, d-1) * dfs(l - ll - 1, m - mm, n - nn, d)) % MOD;
                    add %= MOD;
                }
            }
        }
        dp[l][m][n][d] = add;
        vis[l][m][n][d] = true;
        return add;
    }
}
